### ISSUE TITLE
Mix precision training failed for Pointpillars on Waymo dataset. Got nan output all the time.

### ISSUE THREAD

**Issue by zhanggefan** [open-mmlab/mmdetection3d#299]
Hi, 

I tried training Pointpillars with fp16 on Waymo dataset but got nan output almost every iteration.

After some troubleshooting, I accidentally found out the nan is caused by the inf value at the input point cloud tensor, at the "intensity" column. Waymo uses the raw intensity value and it covers a wide range as mentioned by the following issue:

https://github.com/waymo-research/waymo-open-dataset/issues/226#issue-714104811

I also found that more than half the point clouds in Waymo dataset contain points with intensity values larger than 65504, the maximum value fp16 can hold. As a result overflown happens here:

https://github.com/open-mmlab/mmdetection3d/blob/1b39a483263825ac5a3e57fdf07d380adb7aa16d/mmdet3d/models/detectors/base.py#L46-L49

**Comment by tianweiy** [comment]
You can do a tanh to normalize the intensity value.

**Comment by zhanggefan** [comment]
Hi, @tianweiy 
Thank you for your reply!
That is exactly what I am going to try in the next round of training.
For the latest round, I tired a dirty way and modified the code like this:
```
    @auto_fp16(apply_to=('img', 'points'))
    def forward(self, return_loss=True, **kwargs):
        """Calls either forward_train or forward_test depending on whether
        return_loss=True.

        Note this setting will change the expected inputs. When
        `return_loss=True`, img and img_metas are single-nested (i.e.
        torch.Tensor and list[dict]), and when `resturn_loss=False`, img and
        img_metas should be double nested (i.e.  list[torch.Tensor],
        list[list[dict]]), with the outer list indicating test time
        augmentations.
        """
        # dirty code: clamp the input into the valid range
        for i, p in enumerate(kwargs['points']):
            if isinstance(p, torch.Tensor):
                kwargs['points'][i] = torch.clamp_(p, 
                                                - torch.finfo(p.dtype).max ** 0.75, 
                                                torch.finfo(p.dtype).max ** 0.75)
            elif isinstance(p, list):
                for j, pj in enumerate(p):
                    kwargs['points'][i][j] = torch.clamp_(pj, 
                                                          - torch.finfo(pj.dtype).max ** 0.75, 
                                                          torch.finfo(pj.dtype).max ** 0.75)
        if return_loss:
            return self.forward_train(**kwargs)
        else:
            return self.forward_test(**kwargs)
```
What surprised me is that it actually works, and the result is even a little bit higher than expected(even higher than the official fp32 result):
trained with Waymo Dataset, Load Interval 5, 8 RTX2080Ti, fp16, 24 batchsize in total, fp16 dynamic loss scaling
OBJECT_TYPE_TYPE_VEHICLE_LEVEL_1: [mAP 0.697904] [mAPH 0.691961]
OBJECT_TYPE_TYPE_VEHICLE_LEVEL_2: [mAP 0.621857] [mAPH 0.616351]
OBJECT_TYPE_TYPE_PEDESTRIAN_LEVEL_1: [mAP 0.684974] [mAPH 0.521657]
OBJECT_TYPE_TYPE_PEDESTRIAN_LEVEL_2: [mAP 0.607477] [mAPH 0.460859]
OBJECT_TYPE_TYPE_SIGN_LEVEL_1: [mAP 0] [mAPH 0]
OBJECT_TYPE_TYPE_SIGN_LEVEL_2: [mAP 0] [mAPH 0]
OBJECT_TYPE_TYPE_CYCLIST_LEVEL_1: [mAP 0.57571] [mAPH 0.546155]
OBJECT_TYPE_TYPE_CYCLIST_LEVEL_2: [mAP 0.553777] [mAPH 0.525347]
RANGE_TYPE_VEHICLE_[0, 30)_LEVEL_1: [mAP 0.875814] [mAPH 0.87018]
RANGE_TYPE_VEHICLE_[0, 30)_LEVEL_2: [mAP 0.866379] [mAPH 0.860768]
RANGE_TYPE_VEHICLE_[30, 50)_LEVEL_1: [mAP 0.679541] [mAPH 0.673151]
RANGE_TYPE_VEHICLE_[30, 50)_LEVEL_2: [mAP 0.619451] [mAPH 0.613485]
RANGE_TYPE_VEHICLE_[50, +inf)_LEVEL_1: [mAP 0.455386] [mAPH 0.447845]
RANGE_TYPE_VEHICLE_[50, +inf)_LEVEL_2: [mAP 0.34688] [mAPH 0.340952]
RANGE_TYPE_PEDESTRIAN_[0, 30)_LEVEL_1: [mAP 0.706388] [mAPH 0.548263]
RANGE_TYPE_PEDESTRIAN_[0, 30)_LEVEL_2: [mAP 0.656606] [mAPH 0.508632]
RANGE_TYPE_PEDESTRIAN_[30, 50)_LEVEL_1: [mAP 0.694303] [mAPH 0.52544]
RANGE_TYPE_PEDESTRIAN_[30, 50)_LEVEL_2: [mAP 0.626892] [mAPH 0.472974]
RANGE_TYPE_PEDESTRIAN_[50, +inf)_LEVEL_1: [mAP 0.609291] [mAPH 0.443356]
RANGE_TYPE_PEDESTRIAN_[50, +inf)_LEVEL_2: [mAP 0.480284] [mAPH 0.347692]
RANGE_TYPE_SIGN_[0, 30)_LEVEL_1: [mAP 0] [mAPH 0]
RANGE_TYPE_SIGN_[0, 30)_LEVEL_2: [mAP 0] [mAPH 0]
RANGE_TYPE_SIGN_[30, 50)_LEVEL_1: [mAP 0] [mAPH 0]
RANGE_TYPE_SIGN_[30, 50)_LEVEL_2: [mAP 0] [mAPH 0]
RANGE_TYPE_SIGN_[50, +inf)_LEVEL_1: [mAP 0] [mAPH 0]
RANGE_TYPE_SIGN_[50, +inf)_LEVEL_2: [mAP 0] [mAPH 0]
RANGE_TYPE_CYCLIST_[0, 30)_LEVEL_1: [mAP 0.672245] [mAPH 0.638238]
RANGE_TYPE_CYCLIST_[0, 30)_LEVEL_2: [mAP 0.666969] [mAPH 0.633254]
RANGE_TYPE_CYCLIST_[30, 50)_LEVEL_1: [mAP 0.531476] [mAPH 0.513313]
RANGE_TYPE_CYCLIST_[30, 50)_LEVEL_2: [mAP 0.500889] [mAPH 0.483765]
RANGE_TYPE_CYCLIST_[50, +inf)_LEVEL_1: [mAP 0.411147] [mAPH 0.379818]
RANGE_TYPE_CYCLIST_[50, +inf)_LEVEL_2: [mAP 0.382821] [mAPH 0.353635]

I have not tried training with fp32 yet, nor with intensity-normailized data, but I believe that there is still room for further improvements.

**Comment by tianweiy** [comment]
Looks good. Out of curiosity, how much memory does it take to train pp model using fp16?

**Comment by Tai-Wang** [comment]
@zhanggefan Hi, thanks a lot for your information. We have not tried mixed precision training on Waymo, but the intensity is indeed a problem that is different from the case of other datasets. Congrats that you find a workaround to make it.
As for the performance, actually, it could fluctuate in a certain range when using the Waymo subset to train the model. A little higher or lower, especially considering the overall balance of the performance for different objects, are both normal. So it seems all goes well for your current attempt. Looking forward to more results/observation and PR to explore along this way.

**Comment by zhanggefan** [comment]
When training with fp16 and 3 samples per GPU (which is also the upper limit. 4 samples would cause OOM), PyTorch reported 8907MB GRAM usage, while Nvidia-smi reported more than 10GB GRAM consumed.

pls find the attached log file:
[20210206_001250.txt](https://github.com/open-mmlab/mmdetection3d/files/5938443/20210206_001250.txt)

**Comment by zhanggefan** [comment]
@Tai-Wang Hi, 
I personally think that a more elegant way to address this is to design & register a new Pipeline class and do the normalization in the data preprocessing pipeline.

**Comment by Tai-Wang** [comment]
> @Tai-Wang Hi,
> I personally think that a more elegant way to address this is to design & register a new Pipeline class and do the normalization in the data preprocessing pipeline.

Yes, maybe we can design a pipeline to normalize point features given the target dimension and normalized method, like [Normalize](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/datasets/pipelines/transforms.py#L545) for input image in mmdetection?

**Comment by ZwwWayne** [comment]
Mixed precision training needs tuning of the loss scale. We will migrate to use dynamic scaling strategy based the new release of MMCV after PR https://github.com/open-mmlab/mmcv/pull/791 is merged.

**Comment by tianweiy** [comment]
@zhanggefan  out of curiosity, is fp16 faster for testing? How much speed improvement could we get?

**Comment by zhanggefan** [comment]
> @zhanggefan out of curiosity, is fp16 faster for testing? How much speed improvement could we get?

I have not got a chance to test it yet. But as far as I know, if you are using GPUs with compute capability >= 7.5 (for example A100, V100, or 2080Ti, etc), you may get much better inference performance with fp16.

I use fp16 mainly because it saves GRAMs. I have only one server with 8x2080Ti that does not allow training with batch size > 16. By using fp16 I was able to train with batch size == 24, meanwhile, the total training time is reduced by approximately 1/3.

**Comment by tianweiy** [comment]
Thanks, I tried fp16 with VoxelNet and it is about 25% speedup. I guess the pointpillar may get a larger boost as it uses more standard 2D convolution.

I use fp16 mainly because it saves GRAMs. I have only one server with 8x2080Ti that does not allow training with batch size > 16. By using fp16 I was able to train with batch size == 24, meanwhile, the total training time is reduced by approximately 1/3. <- make sense, thanks!

**Comment by tianweiy** [comment]
@zhanggefan  Additionally, I just found that spconv 1.2.1 (master branch) + cuda 11 uses far less memory than spconv 1.1 (used in CenterPoint / PCdet, not sure what mmdet's spconv is based on).

For instance, I am able to use batch size 4 with fp32 with SECOND's VoxelNet and it consumes about 8-9 GB memory according to nvidia-smi. I guess fp16 should give further memory save.

**Comment by realshijy** [comment]
hi
I trained pointpillar with fp16 on waymo, but got an error: in VFELayer, outputs of norm layer are all nan
https://github.com/open-mmlab/mmdetection3d/blob/master/mmdet3d/models/voxel_encoders/utils.py#L86
Do you have some idea about where this error come from?
Thank you~

**Comment by tianweiy** [comment]
I didn't use fp16 in the end so you may need to ask others

**Comment by realshijy** [comment]
@zhanggefan Hi, may I ask whether you used naivesyncbn? my batch norm always output nan after several iters (

**Comment by realshijy** [comment]
For those who may meet this problem too, I have solved nan problem by switching to dynamic loss scale )

**Comment by zhanggefan** [comment]
@realshijy Hi,
I changed all NaiveSyncBN to the PyTorch 1.6's native SyncBN. 
BUT I did it only for better training speed and definitely, it won't help solve the nan issue.
Like what you have figured out, the problem might be with the grad scaler. congrats that you find the solution!
Btw, did you normalize the intensity value of the input point cloud? What kind of normalizer did you choose?

**Comment by realshijy** [comment]
@zhanggefan Hi, thank you for your reply.
No I didn't. I just added a clamp as you mentioned above. Do you have any suggestion to handle it better?

**Comment by zhanggefan** [comment]
> @zhanggefan Hi, thank you for your reply.
> No I didn't. I just added a clamp as you mentioned above. Do you have any suggestion to handle it better?

Yep, I tried normalizing the input by tanh like what @tianweiy had proposed.
I also found that when train with more iterations, for example using all data rather than only 1/5, using hard clamp still has the possibility of running into nan. Tanh normalizer could solve the nan problem completely. You may have a try:

code:
```
>>> add following code to mmdet3d/datasets/pipelines/transforms_3d.py
@PIPELINES.register_module()
class NormalizeIntensityTanh():
    def __init__(self, intensity_column=3, pre_tanh_gain=1., post_tanh_gain=1.):
        self.intensity_column = intensity_column
        self.pre_tanh_gain = pre_tanh_gain
        self.post_tanh_gain = post_tanh_gain

    def __call__(self, input_dict):
        intensity = input_dict['points'].tensor[:, self.intensity_column]
        intensity = self.post_tanh_gain * (self.pre_tanh_gain * intensity).tanh()
        input_dict['points'].tensor[:, self.intensity_column] = intensity
        input_dict['intensity_norm_cfg'] = dict(
            pre_tanh_gain=self.pre_tanh_gain, 
            post_tanh_gain=self.post_tanh_gain)
        return input_dict

    def __repr__(self):
        repr_str = self.__class__.__name__
        repr_str += f'(pre_tanh_gain={self.pre_tanh_gain}, '
        repr_str += f'post_tanh_gain={self.post_tanh_gain})'
        return repr_str

>>> change the train/test_pipeline configuration in the dataset config file
train_pipeline = [
...
    dict(type='PointsRangeFilter', point_cloud_range=point_cloud_range),
    dict(type='ObjectRangeFilter', point_cloud_range=point_cloud_range),
    dict(type='PointShuffle'),
    dict(type='NormalizeIntensityTanh', intensity_column=3),   # <<<<<<<<<<<<<<<<<<<<<<<<<
    dict(type='DefaultFormatBundle3D', class_names=class_names),
    dict(type='Collect3D', keys=['points', 'gt_bboxes_3d', 'gt_labels_3d'])
]
test_pipeline = [
...
        transforms=[
            dict(
                type='GlobalRotScaleTrans',
                rot_range=[0, 0],
                scale_ratio_range=[1., 1.],
                translation_std=[0, 0, 0]),
            dict(type='RandomFlip3D'),
            dict(
                type='PointsRangeFilter', point_cloud_range=point_cloud_range),
            dict(type='NormalizeIntensityTanh', intensity_column=3),   # <<<<<<<<<<<<<<<<<<<<<<<<<
            dict(
                type='DefaultFormatBundle3D',
                class_names=class_names,
                with_label=False),
            dict(type='Collect3D', keys=['points'])
        ])
]
```

**Comment by realshijy** [comment]
@zhanggefan  thank you very much! I will try it!

**Comment by chyohoo** [comment]
hi, I want to know if I use fp16 training, the internal feature map will also be stored as fp16?
